### ISSUE TITLE
LEGION WARNING "failed to memoize the trace" / low GPU throughput

### ISSUE THREAD

**Issue by roman955b** [flexflow/FlexFlow#232]
I followed the compilation instructions of #231 plus I added the flag `--python-data-loader-type 2 `as a FlexFlow warning hinted me to do.

Testing on a machine with 8xV100 cards, it runs, _but_ I get the below warning about every ~1000 data points (images) when training AlexNet and the throughput times gets down to about 60% of expected numbers.

```
[Metrics] accuracy: 14.315257% (1246 / 8704) sparse_categorical_crossentropy: 2.292534    │
[Metrics] accuracy: 14.330358% (1284 / 8960) sparse_categorical_crossentropy: 2.292337    │
[Metrics] accuracy: 14.257812% (1314 / 9216) sparse_categorical_crossentropy: 2.292219    │
[Metrics] accuracy: 14.210304% (1346 / 9472) sparse_categorical_crossentropy: 2.292063    │
[Metrics] accuracy: 14.113898% (1373 / 9728) sparse_categorical_crossentropy: 2.291939    │
[Metrics] accuracy: 14.002404% (1398 / 9984) sparse_categorical_crossentropy: 2.291959    │
[0 - 7f33c0063700]   20.357225 {4}{runtime}: [warning 1097] LEGION WARNING: WARNING: The r│
untime has failed to memoize the trace more than 5 times, due to the absence of a replayab│
le template. It is highly likely that trace 201 will not be memoized for the rest of execu│
tion. The most recent template was not replayable for the following reason: precondition n│
ot subsumed by postcondition. Please change the mapper to stop making memoization requests│
. (from file /workdisk/FlexFlowMaster/FlexFlow/deps/legion/runtime/legion/legion_trace.cc:│
2025)                                                                                     │
For more information see:                                                                 │
http://legion.stanford.edu/messages/warning_code.html#warning_code_1097                   │
                                                                                          │
epochs 2, ELAPSED TIME = 6.5928s, THROUGHPUT = 3033.59 samples/s
```
This is the AlexNet strategy file that I tested with: [MCMC_strategy.txt](https://github.com/flexflow/FlexFlow/files/7184523/MCMC_strategy.txt). Batch size was set to 256.

**Comment by roman955b** [comment]
@eddy16112 @jiazhihao Any insights to why I see this big penalty to the throughput when running on GPU? What does the warning hint about the  underlying problem?
Thanks!

**Comment by jiazhihao** [comment]
I will look into this later today.

**Comment by ferdiko** [comment]
I've got the same problem. In case that it helps: This goes away if I delete the line ` data_loader.next_batch(ff); ` so the problem could be something with the data loader.

**Comment by ferdiko** [comment]
The warning also goes away if I use data parallelism

**Comment by ferdiko** [comment]
If I move `data_loader.next_batch(ff);` below `runtime->begin_trace(ctx, trace_id);` I don't have the warning and get expected performance -- any ideas?

**Comment by roman955b** [comment]
@jiazhihao What is the impact of @ferdiko 's proposed code change? Is this a valid resolution?

**Comment by roman955b** [comment]
@ferdiko Hi, Did you check the impact of your change any further? With what workloads did you test it with?
Thanks!

**Comment by ferdiko** [comment]
I tested it with CNNs (mixture of experts with CNN submodules). I didn't observe any unexpected behaviour and have been working with this change for the past few days. It's weird however, because I remember that I was already able to run without the warning and without the proposed change.

Here is my (possibly flawed) understanding of what is happening:
In my understanding, the effect of tracing is that the Legion scheduler remembers how it scheduled tasks so it doesn't need to schedule them again in the next iteration (this scheduling can create a bottle neck). In the warning you got, the tracer complains that it can't use the trace that it remembered. I am not entirely sure how including the data loading as part of the trace should resolve that issue, so it would be nice if someone coud confirm. @eddy16112 ? The only reason I could think of is that loading a data batch is scheduled differently if not done according to the trace which also causes later parts to not be replayable. If the data is loaded as remembered in the trace then this is not the case.

**Comment by ferdiko** [comment]
That I didn't get the warning when using data parallelism could also be a hint that loading the data batch is somehow scheduled to different devices when not replaying it from the trace -- and when using data parallelism, this isn't the case anymore, because the task is executed on all devices anyways.

I don't get how loading the batch would be scheduled to different devices though ... (maybe different CPUs are involved or so?? no idea)

**Comment by roman955b** [comment]
In `alexnet.cc` I  moved  `runtime->begin_trace(ctx, 111/*trace_id*/);` to above  `data_loader.next_batch(ff); ` (from line 136 to line 129).

Running the code with this change give me this crash (v100x8):
```
init linear (input): in_dim(9216) out_dim(4096) batch_size(256)
init linear (input): in_dim(4096) out_dim(4096) batch_size(256)
init linear (input): in_dim(4096) out_dim(10) batch_size(256)
[Metrics] accuracy: 8.984375% (23 / 256) sparse_categorical_crossentropy: 2.302586
[0 - 7f9d46181700]    9.600741 {5}{runtime}: [error 364] LEGION ERROR: Trace violation! Operation at index 0 of trace 111 in task top_level (UID 1) was recorded as having 2 regions, but instead has 4 regions in replay.
```

**Comment by ferdiko** [comment]
The line numbering in `https://github.com/flexflow/FlexFlow/blob/master/examples/cpp/AlexNet/alexnet.cc` is different, could you quickly copy paste how you modified it? It seems like you recorded the first trace without the data loading as the first part of the trace (the first forward operator probably takes 2 regions (input, output) and data loader takes 4).

**Comment by roman955b** [comment]
```
for (int iter = 0; iter < iterations; iter++) {
      runtime->begin_trace(ctx, 111/*trace_id*/);
      if (std::strlen(alexnetConfig.dataset_path) == 0) {
        // Only load data once for random input
        if (iter == 0 && epoch == 0)
          data_loader.next_batch(ff);
      } else {
        data_loader.next_batch(ff);
      }
      ff.forward();
      ff.zero_gradients();
      ff.backward();
      ff.update();
      runtime->end_trace(ctx, 111/*trace_id*/);
    }
```

**Comment by ferdiko** [comment]
I see. And you don't provide a data set, right?

Can you try the following:


```
data_loader.next_batch(ff);
for (int iter = 0; iter < iterations; iter++) {
      runtime->begin_trace(ctx, 111/*trace_id*/);
      ff.forward();
      ff.zero_gradients();
      ff.backward();
      ff.update();
      runtime->end_trace(ctx, 111/*trace_id*/);
}
```

**Comment by roman955b** [comment]
OK. That works.
However... Running with 8 GPU's  gives me throughput of  4233.59 samples/s and 4 GPU's gives me 5615.76 samples/s.
Strategies attached:
[MCMC_8gpus.txt](https://github.com/flexflow/FlexFlow/files/7251843/MCMC_8gpus.txt)
[MCMC_4gpus.txt](https://github.com/flexflow/FlexFlow/files/7251844/MCMC_4gpus.txt)

Command line call:
`./build/examples/cpp/AlexNet/alexnet --batch-size 256 -ll:gpu 8 -ll:fsize  8192 -ll:zsize 12192 --budget 0 --import MCMC_8gpus.txt`

**Comment by roman955b** [comment]
If I increase the number of epochs to 10 the numbers change to  8356.08 samples/s (4 GPU's) and  7853.16 samples/s (8 GPU's)

**Comment by ferdiko** [comment]
Okay.

Note that this only works when you don't want to load a new batch every iteration (which is done in the case that you use random data for benchmarking). Otherwise, this should work:
```
for (int iter = 0; iter < iterations; iter++) {
      runtime->begin_trace(ctx, 111/*trace_id*/);
      data_loader.next_batch(ff);
      ff.forward();
      ff.zero_gradients();
      ff.backward();
      ff.update();
      runtime->end_trace(ctx, 111/*trace_id*/);
}
```

About the performance: Try to set `--ll:util 4` and maybe also play around with `--ll:bgwork` (set it to 2 or 4 or so) -- this can help if utility is your bottle neck. I don't know what the best configuration here is and it's unfortunately a trial and error process -- so try some values here. Legion/FlexFlow also has very great profiling tools: https://legion.stanford.edu/profiling/index.html

Using that, you can exactly see devices utilization and which task is executed when.

**Comment by roman955b** [comment]
Thanks for your help.  A few technical questions:
How do I specify a dataset to run with?
How do I execute only a few iterations? I only see the option to control the number of epochs.
The code change discussed in this thread is applied inside the code of workload. How can we achieve the same using the python interface (the alexnet python example doesn't have the for loop from the previous post).

**Comment by ferdiko** [comment]
How many CPUs do you have? If you have more than 1, try to also set `-ll:cpu N`

**Comment by ferdiko** [comment]
1. You can provide a path to a data set using a command line arguement.
2. Few iterations for profiling: `int iterations = data_loader.num_samples / ffConfig.batchSize;` (line 96 in master's `alexnet.cc`). Just set `iterations` to whatever you want.
3. I don't know about python. @eddy16112 @jiazhihao

**Comment by roman955b** [comment]
I ran now with profiling, using the MCMC mapping for 4 and 8 GPU's and I cannot see that anything but the "channel" being (partly) saturated (periodically reaching 100% utilization).  The profiles for 4 and 8 GPU's are close to identical.
Command line arguments: 
`./examples/cpp/AlexNet/alexnet --batch-size 256 -ll:gpu 8 -ll:fsize 8192 -ll:zsize 12192 --budget 0 --import MCMC_8gpus.txt --epochs 1 -d examples/cpp/AlexNet/test_data_0.npz -ll:cpu 10 -ll:util 10 -ll:bgwork 4 -lg:prof 1 -lg:prof_logfile prof_1.gz`
4 gpus: 5011.30 samples/s
8 gpus:  4278.63 samples/s

<img width="668" alt="4gpus_prof" src="https://user-images.githubusercontent.com/84937808/135294647-cfaa59f3-ed10-4fbd-86ac-f6b97f0b0e3f.png">
<img width="668" alt="8gpus_prof" src="https://user-images.githubusercontent.com/84937808/135294667-0fdeb0ad-cad4-46d2-ac2a-e3bd504dc827.png">
[MCMC_8gpus.txt](https://github.com/flexflow/FlexFlow/files/7252933/MCMC_8gpus.txt)
[MCMC_4gpus.txt](https://github.com/flexflow/FlexFlow/files/7252934/MCMC_4gpus.txt)

**Comment by roman955b** [comment]
Strategies of previous post:
[MCMC_8gpus.txt](https://github.com/flexflow/FlexFlow/files/7252963/MCMC_8gpus.txt)
[MCMC_4gpus.txt](https://github.com/flexflow/FlexFlow/files/7252964/MCMC_4gpus.txt)

**Comment by ferdiko** [comment]
It looks like communication is a bottle neck. You can try to increase `-ll:bgwork` for that (which you might have tried already). Apart from that I don't have other ideas ... If increasing `-ll:bgwork` doesn't work, it could be that the problem lies in the model / parallelization strategy and not your configuration. Maybe it's hard in that setting to get rid of the communication bottle neck (increasing batch size could maybe help).

**Comment by ferdiko** [comment]
Things you could check are: On 8 GPUs, is the MCMC strategy faster than data parallelism (just don't set budget and don't do `--import` etc.). If not, then the problem lies with the parallelization strategy -- did you give it enough budget? Did you configure things right? AFAIK FlexFlow doesn't do probing of how much bandwidth your interconnects have etc.

But again, I don't really have good ideas for that

**Comment by roman955b** [comment]
Using the command line:
`./examples/cpp/AlexNet/alexnet --batch-size 256 -ll:gpu 8 -ll:fsize 8192 -ll:zsize 12192  --epochs 1  -ll:cpu 30 -ll:util 10 -ll:bgwork 26 -lg:prof 1 -lg:prof_logfile prof_1.gz `
gives x10 lower throughput (440.04 samples/s)
Seems a little bit too large of a difference to me.

**Comment by ferdiko** [comment]
haha XD. I would try: -ll:cpu 8 -ll:util 4 -ll:bgwork 8 . Something like that. cpu, bgwork (and also util) seem too high to work well. If you can't get on a good track searching in that area, then I also don't have further ideas (except for quickly checking if the parallelization strategy outperforms simple baselines, as mentioned above).

Good luck!

**Comment by roman955b** [comment]
Indeed the optimization of the MCMC reduces the bandwidth bottlenecks. Attached two screenshots from the profiling that indicates how the "`Weight prefetch task`" is reduced from ~140ms down to ~11ms, and the "`SGD Parameter Server Update Task`"  goes down from ~67ms to ~4ms.

So definitely FF does a good job optimizing. However whereas the simulator indicates almost linear positive performance improvement (with MCMC) with increased number of GPU's, actual HW doesn't show this (rather it gives an inverse proportional relationship).

Can others on this forum confirm this observation?

**Comment by santhnm2** [comment]
I'm seeing this warning as well when running a Python Keras MLP model with the MCMC-optimized strategy:
```
[0 - 7f47bc708f80]    4.351344 {4}{runtime}: [warning 1097] LEGION WARNING: WARNING: The runtime has failed to memoize the trace more than 5 times, due to the absence of a replayable template. It is highly likely that trace 101 will not be memoized for the rest of execution. The most recent template was not replayable for the following reason: precondition not subsumed by postcondition. Please change the mapper to stop making memoization requests. (from file /future/u/keshav2/flexflow/deps/legion/runtime/legion/legion_trace.cc:2025)
```
I also do not see this warning when running with vanilla data parallelism, and the throughput with vanilla data parallelism is about 5x better than the throughput with the MCMC strategy. I used the flags `-ll:cpu 8 -ll:util 4 -ll:bgwork 8` as suggested above. Is there a Legion flag we can pass to disable memoization?
